Tangem/Modules/ManageTokens/ManageTokensViewModel.swift

```diff
@@ -10,4 +10,161 @@ import Foundation
+
+    @Injected(\.tokenQuotesRepository) private var tokenQuotesRepository: TokenQuotesRepository
+    @Injected(\.userWalletRepository) private var userWalletRepository: UserWalletRepository
+
+    // I can't use @Published here, because of swiftui redraw perfomance drop
+    var enteredSearchText = CurrentValueSubject<String, Never>("")
+
+    @Published var tokenViewModels: [ManageTokensItemViewModel] = []
+    @Published var isLoading: Bool = true
+    @Published var alert: AlertBinder?
+    @Published var showToast: Bool = false
+
+    // MARK: - Properties
+
+    var hasNextPage: Bool {
+        loader.canFetchMore
+    }
+
+    private lazy var loader = setupListDataLoader()
+
+    private var bag = Set<AnyCancellable>()
+
+    private unowned let coordinator: ManageTokensRoutable
+
+    private var percentFormatter = PercentFormatter()
+    private var balanceFormatter = BalanceFormatter()
+
+    init(coordinator: ManageTokensRoutable) {
+        self.coordinator = coordinator
+
+        bind()
+    }
+
+    func tokenListDidSave() {
+        Analytics.log(.buttonSaveChanges)
+    }
+
+    func onAppear() {
+        Analytics.log(.manageTokensScreenOpened)
+        loader.reset(enteredSearchText.value)
+    }
+
+    func onDisappear() {
+        DispatchQueue.main.async {
+            self.enteredSearchText.value = ""
+        }
+    }
+
+    func fetch() {
+        loader.fetch(enteredSearchText.value)
+    }
+}
+
+// MARK: - Private
+
+private extension ManageTokensViewModel {
+    func bind() {
+        enteredSearchText
+            .dropFirst()
+            .debounce(for: 0.5, scheduler: DispatchQueue.main)
+            .removeDuplicates()
+            .sink { [weak self] string in
+                if !string.isEmpty {
+                    Analytics.log(.tokenSearched)
+                }
+
+                self?.loader.fetch(string)
+            }
+            .store(in: &bag)
+    }
+
+    func setupListDataLoader() -> ListDataLoader {
+        let supportedBlockchains = Set(userWalletRepository.models.map { $0.config.supportedBlockchains }.joined())
+        let loader = ListDataLoader(supportedBlockchains: supportedBlockchains)
+
+        loader.$items
+            .receive(on: DispatchQueue.main)
+            .sink(receiveValue: { [weak self] items in
+                guard let self = self else {
+                    return
+                }
+
+                let alreadyUpdateQuoteCoinIds = tokenViewModels.filter {
+                    $0.priceChangeState != .loading || $0.priceChangeState != .noData
+                }.map { $0.id }
+
+                print(tokenViewModels.map { $0.priceChangeState })
+
+                print(items.count)
+                print(alreadyUpdateQuoteCoinIds.count)
+
+                tokenViewModels = items.compactMap { self.mapToTokenViewModel(coinModel: $0) }
+                updateQuote(by: items.filter { !alreadyUpdateQuoteCoinIds.contains($0.id) }.map { $0.id })
```

Comment:
```suggestion
                let itemsShouldLoadQuote = items.filter { !alreadyUpdateQuoteCoinIds.contains($0.id) }.map { $0.id }
                updateQuote(by: itemsShouldLoadQuote)
```
давай разделим, а то сложно читается, название можно другое

Reply:
Добавил